### ISSUE TITLE
Flow can't distinguish shapes of union types when testing the shapes of properties within the subtypes 

### ISSUE THREAD

**Issue by dclowd9901** [facebook/flow#4328]
You may see this bug in action at https://flow.org/try/#0C4TwDgpgBAYg9nKBeKBvAPgKClYALASwDsBzALigGdgAnYkgGk3QF8BuTTUSWBAG2RosOAEb0K1OqSasOXcNABCAQxqDU2KBCIBXALYUA5Nv2GmOAG7K+OiBXhwm7TtyXKAXus0mDUY7r0zKAB6YKgATQhKBigASSgAayI4AHcoAmBDSlw8aEplPQgYkR1gdLLKPDgdPgATKmACPgERaGURPmhgRGAIZqhNHCHhkdCoADMaOD0c6DApyBpQKABjXJWEzSsbO144Pic5MdiidL0wTsKiYGVGuCJOcZ0iFbvT8aIACjBVAooVNToKAqdwASgoAH4HGhNARxlBPq44PCfjQCgA6ba2dH4ejIJAoLK0eiGUEw4Y0CDAHQ0U6ojFYiAcHAsTSaSnU2lQXTNDgsIA

Hard to explain, but in essence it seems as though when Flow needs to check a property of a subtype of a union in order to determine the subtype, it fails.

**Comment by bendman** [comment]
The minimal case I could think of that produces this issue is this:

```js
type Cup = {| handle: string |};
type Saucer = {| contents: string |};

type Dish = Cup | Saucer;

function fn(param: Dish): ?Cup {
  if (param.handle != null) {
    return param;
  }
  return;
}
```

**Comment by dclowd9901** [comment]
Thanks @bendman

**Comment by rsolomon** [comment]
The original issue is capturing what appears to be a real bug, but I think the example is catching an error for a different reason. The situation itself is somewhat convoluted and involves any attempt at refining a disjoint union that is nested within a property of another refined disjoint union. [Here is a relatively simple example](https://flow.org/try/#0C4TwDgpgBAKhDOwCMUC8UDeAoKVSQC4oByAQWIBosBfAbiy32jkQCY1Mc9wIjiAhSjXqMeUAHIJkHbLgB2fclToMmEqe3SyoCkoOUi1LYAGYZXJnwDCQ+VKKTESA6rHGALOdyWSNqncQHDRdRSChSABsI43gOYxQAH1gNKCTjMzSpdxEAMwBXOQBjYABLAHs5PCkACmAiDwBKIgA3MpKAE05cQor4MoiIADoIsoBzauIAdTKygAsZgEJKPAbaKBoGfKLSiqq2WqJI6Kl4JqhWjq6oEpyoWsG1VCffYigAMje8Qbkpb7RngTEBpXbw1YCrda4agbLBAA).

The two examples given so far are failing not due to this bug, but because they are attempting to refine a disjoint union using a property name that is not guaranteed to be available on all union branches. This scenario is outlined in [the documentation](https://flow.org/en/docs/types/unions/#toc-disjoint-unions): "Disjoint unions require you to use a single property to distinguish each object type. You cannot distinguish two different objects by different properties."

**Comment by joefraley** [comment]
@rsolomon what about this workaround from the documentation? It comes right after the section you refer to. It claims that **exact object types** circumvent this constraint by forcing a property conflict between the disjointed types:

> However, to get around this you could use **exact object types**.
```javascript
// @flow
type Success = {| success: true, value: boolean |};
type Failed  = {| error: true, message: string |};

type Response = Success | Failed;

function handleResponse(response: Response) {
  if (response.success) { // note that (response.success != null) produces the type error here, why?
    var value: boolean = response.value;
  } else {
    var message: string = response.message;
  }
}
```
> With exact object types, we cannot have additional properties, so the objects conflict with one another and we are able to distinguish which is which.

How is this semantically different from @bendman's case?

**Comment by joefraley** [comment]
@rsolomon sorry, re-reading this thread, i think what we're all saying is:

```javascript
// @flow
type Success = {| success: true |};
type Failed  = {| error: true |};
type Response = Success | Failed;

function handleResponse(response: Response) {
  if (response.success) // this
  if (response.success != null) // should behave the same as this
  if (response.success != undefined) // should behave the same as this
  if (String(response.success)) // should behave the same as this
  // etc
}
```

**Comment by rsolomon** [comment]
Fair enough, I should have kept reading! I think we are maybe capturing 2 different issues here:

1. What you are talking about, regarding refinement of a union based on a negative check.

2. Inability to refine a top level disjoint union when attempting to refine based on a property that is also a disjoint union.

**Comment by bendman** [comment]
I'm wondering if there's consensus on whether this should be fixed.  It would appear that this reveals many related issues, in that (almost?) any check other than comparing an always `true` field using `if (object.field)` will throw an error.

[Some examples](https://flow.org/try/#0C4TwDgpgBAygrgYwRAziqBeKBvAPlYAJzggGEALCBAawC4DiIAaKAMwEMAbFMym+jt2ZQARgHsxnACrheVOqImcI7AHZRcAXwDcAKFCQoAMXYBLZQBMomHPgiFCYwvSIkNO-bKgAlVGDGqPDbwSKjo+CbmEBZ6ugD0cVAA6k7UpqoA5rq6rHCqCMCmAVAIfNRGTlLEwOQgVSQAFIR+ATz0vij+gRAAlPQA-CHIaDi61qasUE0t3QB0rnI0PVDNwHCE6s2drRB6mtkJxmac6Vk5eQVF6qXyFYQAogCOcFz1ENPb3e0zPH1Qg4hhuhsGMoBMplsujx5owKPJMFgFstVutNj9drp9ud8oVijcaHd7gAPdgFTh1RgfKEQb6fX4DIZhUbjSZUnYwkhwmgIxGMZEQNYbFbovbZXI4q4lMqEklkkAmIRsr4+dF-AGhEYglkQ9GzQQ8LnUHlsLi-YWCtF0jFY8WXPHSyqyJVtFVWtWMzWg8ENAwQMSTSHs8SSGSQQ3GgDkweUagj-ItwqteyAA) are below, although there are many more.  I haven't yet looked into the cause of the issues, but I would expect all of these to work:

```js
type Success = {| trueCheck: true, falseCheck: false, boolTypeCheck: boolean |};
type Failed  = {| error: true |};
type Response = Success | Failed;

// Working

function checkForTruthyTrue(response: Response): ?Success {
  if (response.trueCheck) return response;
}

// Failing

function checkForEqualTrue(response: Response): ?Success {
  if (response.trueCheck == true) return response;
}

function checkForExactlyTrue(response: Response): ?Success {
  if (response.trueCheck === true) return response;
}

function checkForExactlyFalse(response: Response): ?Success {
  if (response.falseCheck === false) return response;
}

function checkForType(response: Response): ?Success {
  if (typeof response.boolTypeCheck === 'boolean') return response;
}
```

One note is that some of this could get into [equality oddities](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Equality_comparisons_and_sameness#Loose_equality_using), but strict types help minimize those cases.

**Comment by joefraley** [comment]
yeah, is there another open issue about this? i'm shocked if this is the first mention. this kind of checking is the most natural way to handle disparate inputs in js

**Comment by bendman** [comment]
For the record, the TypeScript version works, but you must explicitly state your types in the conditional, because it doesn't infer anything.

```typescript
interface Success { trueCheck: true; falseCheck: false; boolTypeCheck: boolean; }
interface Failed { error: true; falseCheck: true; boolTypeCheck: string }
type Resp = Success | Failed;

// Working
function checkForTruthyTrue(response: Resp): Success | undefined {
    if ((<Success>response).trueCheck) return <Success>response;
}

function checkForEqualTrue(response: Resp): Success | undefined {
    if ((<Success>response).trueCheck == true) return <Success>response;
}

function checkForExactlyTrue(response: Resp): Success | undefined {
  if ((<Success>response).trueCheck === true) return <Success>response;
}

function checkForExactlyFalse(response: Resp): Success | undefined {
  if ((<Success>response).falseCheck === false) return <Success>response;
}

function checkForType(response: Resp): Success | undefined {
  if (typeof (<Success>response).boolTypeCheck === 'boolean') return <Success>response;
}
```

**Comment by joefraley** [comment]
i think the relevant test code is here:

https://github.com/facebook/flow/blob/3ea841038418f680b0985f8a6084e507b04a484f/tests/exact/prop_test.js#L19

**Comment by joefraley** [comment]
i started poking around the `hack` codebase for these prop checks, but it's a pretty complex system. i think starting on this would require some guidance from a 🐫

**Comment by bendman** [comment]
So, to clarify things, doing a check for `if (object.property)` will refine the scope when _Exact_ object types are used.  This seems to serve the same purpose as TypeScript's `(<Type>variable)` assertions.

@dclowd9901's original issue can be solved in a few equivalent ways:

```js
type Foo = {
  thing: 'thing',
};

type Fool = {|
  bing: 'bing',
|};

type Bar = {
  enum: 'enum',
  value: Foo,
};

type Baz = {
  enum: 'enum',
  // You could also forego the pipes (`|`) in Fool above by using `$Exact<Fool>` here
  value: Fool,
};

// In implementation

function fn(param: Bar | Baz): ?Foo {
  // Note the initial check for `param.value.thing`, refining the type
  if (param.value.thing && param.value.thing === 'thing') {
    return param.value;
  }
  
  return null;
}
```

This works in this case (and [in the negative](https://flow.org/try/#0C4TwDgpgBAYg9nKBeKBvAPgKClYALASwDsBzALigHJ9iTKAaTdAXwG5NNRJYEAbZNFhwAjWhUqjSDJmw5doAIQCGAJwEZsUCEQCuAW3Hb90nADclvHRArw4jFu07hFSgF7qhW3QapG9JqHNLax44XntZTAB6KKgASSIoAj0wXgg9bWAlYAI4Ig4AMx0iAGMcvKgCogAKMFUlH2U1dChlVwBKCgB+W35UTQICqGqAQjqVBoA6IKtJmlJ2tE0cFQhgHRVE8amZiHYcZk1NVfXNqF1eXnZmIA)) but it wouldn't work for some cases, like [typeof checks](https://flow.org/try/#0C4TwDgpgBAYg9nKBeKBvAPgKClYALASwDsBzALiiIFcBbAIwgCcAaTdAXwG5NNRJYEAG2RosOfMXJQAzsEaTWOOpIqz5pVh269w0AEIBDRiIzYoEajQoByC7WuKoANwOCqECvDiauPPvoMALxMxc0sbOxoHMxc3DwE4QR9tAHoUqABJIigCGjBBCBoLYANgAjgiHgAzKiIAYzKKqCqiAAowIwMrKENjdB6ggEoKAH4vNDMCKqhW-zhpjsYugDpY92WJUmQUa0sGRmtBiZwcRghgKkZsxZW1iG4cdjMzM4uryipBQW52IA).

**Comment by chasingmaxwell** [comment]
Maybe it belongs in another issue, but I'm seeing that when the distinguishing property is nested, there are errors, but only when you assert the type of the top-level object in the union. Referencing it's child properties is fine. Here's a simple example:

```JavaScript
// @flow
type Success = { nested: { success: true, value: boolean } };
type Failed  = { nested: { success: false, error: string } };

type Response = Success | Failed;

function handleResponse(response: Response) {
  if (response.nested && response.nested.success) {
    var value: boolean = response.nested.value; // Works!
  } else {
    var error: string = response.nested.error; // Works!
    acceptFailedNested(response.nested); // Works!
	acceptFailed(response); // Fails :(
  }
}

function acceptFailedNested(failedNested: { success: false, error: string }) {
  console.log(failedNested.error);
}

function acceptFailed(failure: Failed) {
  console.log(failure.nested.error);
}
```
[Try it here](https://flow.org/try/#0PTAEAEDMBsHsHcBQAXAngBwKagMoFcBjAzAZxNAF5QBvUAO1OUwBMAuG0Ew4s95AJzyYANKABuAQ2hD2AI1ixomCXVABfdQG4UGbADEJASyXNQlDgxJM2HLkVIl2kKSRGhM-frH7sr-Q3QA5upaiDpYoABKpOiwdK7m+PZkoAA+oAbGLNqIkHh0BMiGcaAAFirMStEksfGYABT8MXGu7NW1rgCUNIhmhpCgjc11AHSW1qAAZJOgTTUtmGOMLCN2PCTd1L1m4hL8u9KYcgpKKuZzHYvjK5KHmqAgoADq3gDWJACE2xqY0AlbO12+w8Xh8nAEAWCVAuCyWVhWIO890eL347y+gIk9nQyEyJgAcstmEN5qNrsxOsiwKj0YgAJBY4g4vEsEmXSkPMB48iserfRBqMJ5ApFEqMzDMowEon1ZxZZiE+E2WhrBxOFxuRFgvyQ9SbbYEFqKRZwQKyqUsRXWEZaykCoX5QrFVTiyXy83GPBNdgsik9MyG+LGkamj3SJpw6227RqIA)

I also tried @bendman's suggestion of using exact types. That did not seem to make any difference.
